common/determined_common/schemas/schema.py

```diff
@@ -0,0 +1,431 @@
+            ):
+                ## This is effectively what happens magically due to @auto_init
+                # if a is not None:
+                #     self.a = a
+                # if b is not None:
+                #     self.b = b
+                pass
+
+    The simpler strategy would be to skip the annotations at the Class level and just set values
+    in __init__.  However, using @auto_init has several benefits:
+
+      - By relying on class annotations for default values, you can always call thing.a, but you
+        can also use `"a" in vars(thing)` to know if the value was set explicitly or not.
+
+      - The annotations are easily recognized by type-aware systems for linting or tab-completion.
+
+      - Listing out the types in the signature of __init__() is not actually necessary (it could
+        be inferred from the annotations) but for type-awarness and tab-completion systems it is
+        necessary.  (side note: you don't need explicit __init__ definitions for @dataclass
+        classes when working with mypy, but that's because mypy special-cases them.)
+
+      - Given the previous point, the annotations and the signature can easily be kept in perfect
+        sync with each other, but enforcing the synchronization between the __init__ signature and
+        the body of __init__ would be extremely difficult in a large body of evolving configs.
+
+    Also see the AutoInit metaclass for an *even easier* way to get the benefits of @auto_init.
+    """
+
+    old_sig = inspect.signature(old_init)
+
+    @functools.wraps(old_init)
+    def set_all_attrs(self: T, *args: list, **kwargs: dict) -> None:
+        if args:
+            raise TypeError("only use keyword arguments")
+
+        kw = dict(old_sig.bind(self, *args, **kwargs).arguments)
+        del kw["self"]
+        for k, v in kw.items():
+            setattr(self, k, v)
+
+        # Always call the old __init__ in case there is anything useful in there.
+        old_init(self, **kwargs)
+
+    return set_all_attrs
+
+
+class AutoInit(type):
+    """
+    AutoInit is a metaclass (it subclasses type) and it is invoked when a type is created, rather
+    than an instance of a type is created.  That means, it's invoked at the time a class is DEFINED
+    rather than when it is INSTANTIATED.  That lets us do things like apply the @auto_init
+    decorator on every single subclass of SchemaBase (whose metaclass=AutoInit) and call it a day.
+    """
+
+    def __new__(cls: type, name: str, bases: Tuple, dct: dict) -> Any:
+        # Nothing to do on either the SchemaBase class or if there is no __init__ function.
+        if name == "SchemaBase" or "__init__" not in dct:
+            return super().__new__(cls, name, bases, dct)  # type: ignore
+
+        dct["__init__"] = auto_init(dct["__init__"])
+
+        return super().__new__(cls, name, bases, dct)  # type: ignore
+
+
+T = TypeVar("T", bound="SchemaBase")
+
+
+class SchemaBase(metaclass=AutoInit):
```

Comment:
I thought you were going to remove this metaclass?

Reply:
I will take care of this in a follow-up PR.